servant-checked-exceptions/example/CustomEnvelope.hs

```diff
@@ -0,0 +1,279 @@
+
+
+-- | Dummy FooBar result type
+-- When serialised, it is contained inside a "foobar" field.
+data FooBar = FooBar String
+
+instance ToJSON FooBar where
+  toJSON (FooBar s) = object $ ["foobar" .= s]
+
+instance FromJSON FooBar where
+  parseJSON = withObject "FooBar" $ \o -> FooBar <$> o .: "foobar"
+
+-- | Docs for FooBar as response
+instance ToSample FooBar where
+  toSamples Proxy = [("A successful FooBar response", FooBar "foo: 42")]
+
+------------
+-- Errors --
+------------
+
+-- | Dummy errors: these serialise in an "error" field.
+data Err1 = Err1 deriving Show
+
+instance ToJSON Err1 where
+  toJSON Err1 = object $ [ "error" .= show Err1]
+
+instance FromJSON Err1 where
+  parseJSON = withObject "Err1" $ \o -> do
+    e <- o .: "error"
+    guard $ e == show Err1
+    pure Err1
+
+-- | Error status
+instance ErrStatus Err1 where
+  toErrStatus _ = status400
+
+-- | Sample error
+instance ToSample Err1 where
+  toSamples Proxy = [("Error nr. 1", Err1)]
+
+
+data Err2 = Err2 deriving Show
+
+instance ToJSON Err2 where
+  toJSON Err2 = object $ ["error" .= show Err2]
+
+instance FromJSON Err2 where
+  parseJSON = withObject "Err2" $ \o -> do
+    e <- o .: "error"
+    guard $ e == show Err2
+    pure Err2
+
+instance ErrStatus Err2 where
+  toErrStatus _ = status400
+
+instance ToSample Err2 where
+  toSamples Proxy = [("Error nr. 2", Err2)]
+
+---------
+-- API --
+---------
+
+-- | This is our main 'Api' type. It defines two simple routes: one throws
+-- errors, the second throws no errors. A custom envelope is used for both.
+-- We will create a server, a client, and documentation for this api.
+-- To understand what is different, compare with the modules 'Api', 'Client' and 'Server'.
+type Api = "api" :> ( FooApi :<|> BarApi)
+
+type FooApi = 
+  Capture "foo" Foo :>
+  Throws' FlatEnvelope Err1 :>
+  Throws' FlatEnvelope Err2 :>
+  Post '[JSON] FooBar
+
+type BarApi = 
+  Capture "bar" Bar :>
+  NoThrow' FlatEnvelope :>
+  Post '[JSON] FooBar
+
+------------
+-- Server --
+------------
+
+-- | This is our server root for the 'ServerT' for 'Api'.  We have two handlers.
+serverRoot :: ServerT Api Handler
+serverRoot = fooHandler :<|> barHandler
+
+-- | This is the handler for 'FooApi'.
+-- If the integer contained in 'Foo' is '1' or '2', we respectively throw an
+-- 'Err1' or 'Err2'. Otherwise, we return success.
+-- We use standard envelopes, and at the outer layer wrap it with a custom
+-- envelope. This allows to reuse existing envelope functions and 'EnvelopeT',
+-- but still have custom JSON instances. You can of course define your own
+-- custom envelope type which does not use 'Envelope' internally.
+fooHandler :: Foo -> Handler (FlatEnvelope '[Err1, Err2] FooBar)
+fooHandler (Foo 1) = fmap FlatEnvelope . pureErrEnvelope $ Err1
+fooHandler (Foo 2) = fmap FlatEnvelope . pureErrEnvelope $ Err2
+fooHandler (Foo n) = fmap FlatEnvelope . pureSuccEnvelope . FooBar $ "foo: " <> show n
```

Comment:
Not being able to directly use `pureErrEnvelope` and `pureSuccEnvelope` when using a different envelope type is pretty inconvenient.

I guess all of the functions exported by `Servant.Checked.Exceptions.Envelope` would have to be wrapped?

Reply:
Not strictly necessary. My current implementation uses it like this: I program using normal `Envelope`s (`EnvelopeT`), and only right before returning the return value in the endpoint handler, I wrap it in a `FlatEnvelope`. Combined with an existing helper function, this makes the conversion of my server-side code only a oneliner.